rsciio/jobin_yvon/api.py

```diff
@@ -0,0 +1,819 @@
+            ## reshape depends on available axes
+            if self._has_nav2:
+                if self._has_nav1:
+                    self.data = np.reshape(
+                        self.data, (self._nav2_size, self._nav1_size, num_cols)
+                    )
+                else:
+                    self.data = np.reshape(
+                        self.data, (self._nav2_size, num_cols)
+                    )  # pragma: no cover
+            elif self._has_nav1 and not self._has_nav2:
+                self.data = np.reshape(self.data, (self._nav1_size, num_cols))
+
+    ## account for missing keys in original metadata (try/except)
+    def map_metadata(self):
+        """Maps original_metadata to metadata dictionary."""
+        self.metadata = dict()
+        if "General" not in self.metadata:
+            self.metadata["General"] = {}
+        if "Signal" not in self.metadata:
+            self.metadata["Signal"] = {}
+        if "Sample" not in self.metadata:
+            self.metadata["Sample"] = {}
+        if "Acquisition_instrument" not in self.metadata:
+            self.metadata["Acquisition_instrument"] = {
+                "Laser": {"Filter": {}, "Polarizer": {}},
+                "Spectrometer": {"Grating": {}, "Polarizer": {}},
+                "Detector": {"processing": {}},
+                "Spectral_image": {},
+            }
+
+        self.metadata["General"]["title"] = self._title
+        self.metadata["General"]["original_filename"] = self._file_path.name
+        self._set_metadata(
+            self.metadata["General"],
+            "notes",
+            self.original_metadata["file_information"],
+            "Remark",
+        )
+        try:
+            date, time = self.original_metadata["date"]["Acquired"].split(" ")
+            self.metadata["General"]["date"] = date
+            self.metadata["General"]["time"] = time
+        except KeyError:
+            pass  # pragma: no cover
+
+        try:
+            intensity_axis = self.original_metadata["experimental_setup"]["signal type"]
+            intensity_units = self.original_metadata["experimental_setup"][
+                "signal units"
+            ]
+        except KeyError:
+            pass  # pragma: no cover
+        else:
+            if intensity_axis == "Intens":
+                intensity_axis = "Intensity"
+            if intensity_units == "Cnt/sec":
+                intensity_units = "Counts/s"
+            if intensity_units == "Cnt":
+                intensity_units = "Counts"
+            self.metadata["Signal"][
+                "quantity"
+            ] = f"{intensity_axis} ({intensity_units})"
+
+        self.metadata["Signal"]["signal_type"] = self._signal_type
+        self.metadata["Signal"]["signal_dimension"] = 1
+
+        self._set_metadata(
+            self.metadata["Sample"],
+            "description",
+            self.original_metadata["file_information"],
+            "Sample",
+        )
```

Comment:
I agree. 

Mostly, I dislike the `_set_metadata` paradigm, because in essence all it does is assigning in a rather obscure way:
`self.metadata["Sample"]["description"] = self.original_metadata["file_information"]["Sample"]`

I would rather do the assignment explicitly with None values if keys don't exist, and then remove the None.

e.g.
```python
Laser = {}
Laser['wavelength'] = self.original_metadata["experimental_setup"].get("Laser (nm)")
Laser["objective_magnification"] = self.original_metadata["experimental_setup"].get("Objective")
#etc etc
self.metadata["Acquisition_instrument"]["Laser"] = {k: v for k, v in Laser.items() if v is not None}
```

Reply:
I like the idea of using .get() and removing None values. This is now implemented.